### PR TITLE
Dispersive Laser

### PR DESCRIPTION
**Pull request by fabidie**
The new profile "DispersiveLaser" allows to simulate laser pulses with dispersive properties like AD, SD, GDD and TOD. One can set these parameters in DispersiveLaser.def. The incident E field is calculated via Fourier transformation and the B field by the already existing "approximateB" method in PiconGPU, using SVEA. 
There is another possibility implemented to calculate the incident B field, also via DFT, but it is still buggy and leads to wrong results, that is why it is not in use yet (Klaus said to me he will fix this).

Here some proof that this profile works (with approximateB als incident Field):

For the following simulation output I used lambda = 0.8µm, focus position in propagation direction at 2e-5m, w0 = 5µm / 1.17741 and tau = 10fs / 1.17741

- setting AD to -2.45e-16 rad*s, which should result in a pulse front tilt of around 30° in focus
AD_30_focus
AD_30_zR

All other dispersion parameters have been tested too and lead to satisfying results aswell.

- comparing GaussianBeam profile with DispersiveLaser profile whilest setting all dispersion parameters to zero leads to similar results:
vgl_DL_GB_all
vgl_DL_GB_slice

### CODE REVIEW

File: include/picongpu/fields/incidentField/profiles/DispersiveLaser.def [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                        
+                        /** SD: spatial dispersion in focus
+                         * = dx0 / dΩ
+                         *
+                         * unit: m * s
+                         */
+                        static constexpr float_64 SD_SI = 0.0;
+                        
+                        /** AD: angular dispersion in focus
+                         * dθ/dΩ, e.g. tan(α_tilt)/Ω0
```
Comment (finnolec) [inline include/picongpu/fields/incidentField/profiles/DispersiveLaser.def:67]: I just skimmed over the code, but I noticed the greek letters like Ω,θ and α everywhere in the comments. I don't think we want to use unicode characters in PIConGPU. So `* d theta / d Omega, e.g. tan(alpha_tilt)/(Omega_0)` would probably better here (and in the other places with the greek letters).

Comment (sbastrakov) [reply include/picongpu/fields/incidentField/profiles/DispersiveLaser.def:67]: Yes, please use only ASCII characters in the source code. Using TeX is possible in readthedocs, there it would also be rendered properly.

**Comment by psychocoderHPC** [comment]
Please run once `clang-format-12` to fix the formatting issues.

https://picongpu.readthedocs.io/en/latest/dev/docs/COMMIT.html?highlight=clang-format#picongpu-commit-rulez

File: include/picongpu/fields/incidentField/profiles/GaussianBeam.hpp [hunk 2]
```diff
@@ -251,7 +253,9 @@ namespace picongpu
                             // inverse radius of curvature of the beam's  wavefronts
                             auto const R_inv = -focusPos / (Unitless::R * Unitless::R + focusPos * focusPos);
                             // the Gouy phase shift
-                            auto const xi = math::atan(-focusPos / Unitless::R);
+                            auto xi = math::atan(-focusPos / Unitless::R);
+                            if(simDim == DIM2)
+                                xi *= 0.5_X;
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/GaussianBeam.hpp:258]: This looks like a bug fix for the already existing profiles.
We should separate these changes to be able to write a clean changelog.

File: include/picongpu/fields/incidentField/profiles/BaseParam.def [hunk 3]
```diff
@@ -117,9 +117,9 @@ namespace picongpu
                      *
                      * @{
                      */
-                    static constexpr float_64 FOCUS_POSITION_X_SI = 8.0e-5;
-                    static constexpr float_64 FOCUS_POSITION_Y_SI = 5.0e-5;
-                    static constexpr float_64 FOCUS_POSITION_Z_SI = 8.0e-5;
+                    static constexpr float_64 FOCUS_POSITION_X_SI = 5.0e-5;
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/BaseParam.def:120]: Why are these lines changed? Please revert the all changes in this file.

File: include/picongpu/fields/incidentField/Functors.hpp [hunk 4]
```diff
@@ -353,11 +353,14 @@ namespace picongpu
                             // In principle 1 ulp should work, but just to be safe against changes in unit system
                             constexpr double ulp = 4.0;
                             constexpr double eps = std::numeric_limits<double>::epsilon();
-                            bool const isMatchingUnit = (std::fabs(unitField[axis] - UNIT_EFIELD) <= eps * ulp);
+                            bool const isMatchingUnit
+                                = (std::fabs(unitField[axis] - UNIT_EFIELD) <= eps * UNIT_EFIELD * ulp);
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/Functors.hpp:357]: This fix should be pushed as a separate PR.

File: include/picongpu/param/incidentField.param [hunk 5]
```diff
@@ -32,6 +32,7 @@
  *  - profiles::PulseFrontTilt<>      : Gaussian beam with tilted pulse front with given parameters
  *  - profiles::Wavepacket<>          : wavepacket with Gaussian spatial and temporal envelope profile with given
  * parameters
+ *  - profiles::DispersiveLaser<>     : dispersive Gaussian Beam with possibilities to set SD, AD, GDD and TOD
```
Comment (psychocoderHPC) [inline include/picongpu/param/incidentField.param:35]: Could it be that comment for the spectral laser profile is missing?

File: include/picongpu/fields/incidentField/profiles/SpectralLaser.hpp [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+                            // central frequency Unitless::w
+                            // gaussian envelope in frequency domain
+                            // using pulse length - bandwidth product: sigma_Omega = sqrt(2)/tau
+                            float_X norm = math::sqrt(pmacc::math::Pi<float_X>::halfValue * 2.0_X) * Unitless::PULSE_LENGTH * 2.0_X;
+                            float_X eps = norm * math::exp(-(Omega - Unitless::w) * (Omega - Unitless::w) * Unitless::PULSE_LENGTH * Unitless::PULSE_LENGTH);
+                            
+                            float_X amp = eps * math::exp(-r2OverW2);
+
+                            // distinguish between dimensions
+                            if(simDim == DIM2)
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/SpectralLaser.hpp:164]: should be `if constexpr(simDim == DIM2)` this will enforce the compile to evaluate this part at compile time.

File: include/picongpu/fields/incidentField/profiles/SpectralLaser.hpp [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+                                amp *= math::sqrt(Unitless::W0 / waist);
+                            else if(simDim == DIM3)
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/SpectralLaser.hpp:166]: else if constexpr(simDim == DIM3)

File: include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+
+                            // Center of a frequency's spatial distribution
+                            float_X center = Unitless::SD * (Omega - Unitless::w)
+                                           + SPEED_OF_LIGHT * alpha * focusPos / (Unitless::W0 * Unitless::w);
+
+                            // gaussian envelope in frequency domain
+                            float_X mag = math::exp(-(Omega - Unitless::w) * (Omega - Unitless::w) * Unitless::PULSE_LENGTH * Unitless::PULSE_LENGTH);
+
+                            // distinguish between dimensions
+                            if(simDim == DIM2)
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp:167]: `if constexpr(simDim == DIM2)`

File: include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+                            {
+                                // pos has just two entries: pos[0] as propagation direction and pos[1] as transversal direction
+                                mag *= math::exp(-(pos[1]-center) * (pos[1]-center) / waist / waist);
+                                mag *= math::sqrt(Unitless::W0 / waist);
+                            }
+                            else if(simDim == DIM3)
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp:173]: else if constexpr(simDim == DIM3)

File: include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+                            float_X const timeDelay = mue + focusPos / SPEED_OF_LIGHT;
+
+                            float_X phase = - Omega * focusPos / SPEED_OF_LIGHT
+                                            + 0.5_X * Unitless::GDD * (Omega - Unitless::w) * (Omega - Unitless::w)
+                                            + Unitless::TOD / 6.0_X * (Omega - Unitless::w) * (Omega - Unitless::w) * (Omega - Unitless::w)
+                                            + phaseShift + Unitless::LASER_PHASE
+                                            + Omega * timeDelay;
+
+                            // distinguish between dimensions
+                            if(simDim == DIM2)
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp:232]: if constexpr(simDim == DIM2)

File: include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+                            {
+                                phase += (pos[1] - center) * (pos[1] - center) * Omega * 0.5_X * R_inv / SPEED_OF_LIGHT;
+                                phase -= alpha * pos[1] / Unitless::W0;
+                                phase += alpha * alpha / 4.0_X * focusPos / Unitless::R;
+                                phase -= 0.5_X * xi;
+                            }
+                            else if(simDim == DIM3)
```
Comment (psychocoderHPC) [inline include/picongpu/fields/incidentField/profiles/DispersiveLaser.hpp:239]: `else if constexpr(simDim == DIM3)`

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]
Very nice PR, I added some comments.

File: include/picongpu/fields/incidentField/Functors.hpp (hunk 4 above)
Comment (sbastrakov) [reply include/picongpu/fields/incidentField/Functors.hpp:357]: It is already fixed in `dev` now.

File: include/picongpu/fields/incidentField/profiles/BaseParam.def (hunk 3 above)
Comment (sbastrakov) [reply include/picongpu/fields/incidentField/profiles/BaseParam.def:120]: Yes, I suspect the testing for this PR was done using the default param structure and so these parameters were adjusted. But this is not a workflow PIConGPU's user should be doing, so these default values should not make any difference, they are merely for illustration.

File: include/picongpu/fields/incidentField/profiles/GaussianBeam.hpp (hunk 2 above)
Comment (sbastrakov) [reply include/picongpu/fields/incidentField/profiles/GaussianBeam.hpp:258]: It seems to be a fix of #4249 . Indeed, should be in a separate PR.

File: include/picongpu/fields/incidentField/Functors.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/picongpu/fields/incidentField/Functors.hpp:357]: ok in this case a rebase is required.

File: include/picongpu/fields/incidentField/profiles/GaussianBeam.hpp (hunk 2 above)
Comment (steindev) [reply include/picongpu/fields/incidentField/profiles/GaussianBeam.hpp:258]: See #4301

File: include/picongpu/param/incidentField.param (hunk 5 above)
Comment (steindev) [reply include/picongpu/param/incidentField.param:35]: The SpectralLaser profile should not be part of this PR anyways. It is just a (unnecessary) specialization of the DispersiveLaser.
I will completely remove it from this PR.

File: include/picongpu/fields/incidentField/profiles/BaseParam.def (hunk 3 above)
Comment (steindev) [reply include/picongpu/fields/incidentField/profiles/BaseParam.def:120]: Reverted.

File: include/picongpu/fields/incidentField/Functors.hpp (hunk 4 above)
Comment (steindev) [reply include/picongpu/fields/incidentField/Functors.hpp:357]: Done.

File: include/picongpu/fields/incidentField/profiles/SpectralLaser.hpp (hunk 6 above)
Comment (steindev) [reply include/picongpu/fields/incidentField/profiles/SpectralLaser.hpp:164]: Done

File: include/picongpu/fields/incidentField/profiles/DispersiveLaser.def (hunk 1 above)
Comment (steindev) [reply include/picongpu/fields/incidentField/profiles/DispersiveLaser.def:67]: I hope I catched all.

**Comment by steindev** [comment]
> Please run once `clang-format-12` to fix the formatting issues.
> 
> https://picongpu.readthedocs.io/en/latest/dev/docs/COMMIT.html?highlight=clang-format#picongpu-commit-rulez

Done.

**Comment by steindev** [comment]
Superseded by #4302.
